### PR TITLE
Write IAM credentials to credentials file and show error if profile already exists

### PR DESCRIPTION
**Pull request by rli**
image

Highlights
* Expand `ConfigFileWriter` to a more general facade to accommodate better testing in `sso-session` logic
* credentials written to `credentials` instead of `config` per previous discussion offline
* Add missing scopes test for IdC case due to regression caused by previous commit

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [ ] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [ ] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ConfigFilesFacade.kt [hunk 1]
```diff
@@ -164,17 +102,76 @@ object DefaultConfigFileWriter : ConfigFileWriter {
...
+        appendSection(configPath, sectionName, profile)
+
+    private fun appendSection(path: Path, sectionName: String, profile: Profile) {
+        val isConfigFile = path.fileName.toString() != "credentials"
+        if (sectionName == "sso-session" && !isConfigFile) {
+            error("sso-session is only allowed in 'config'")
+        }
+
+        // behavior is different between config/credentials
+        val sectionTitle = if (!isConfigFile || profile.name().trim() == "default") {
```
Comment (manodnyab) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ConfigFilesFacade.kt:157]: nit: For reference could we add a comment with an example of the structure of this section

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/CreateOrUpdateCredentialProfilesAction.kt [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import org.jetbrains.annotations.TestOnly
+import software.amazon.awssdk.profiles.ProfileFileLocation
+import software.aws.toolkits.core.utils.exists
+import software.aws.toolkits.resources.message
+import software.aws.toolkits.telemetry.AwsTelemetry
+import java.nio.file.Path
+
+class CreateOrUpdateCredentialProfilesAction @TestOnly constructor(
+    private val writer: ConfigFilesFacade
+) : AnAction(message("configure.toolkit.upsert_credentials.action")), DumbAware {
```
Comment (manodnyab) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/CreateOrUpdateCredentialProfilesAction.kt:28]: nit: DumbAwareAction?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/CreateOrUpdateCredentialProfilesAction.kt [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        // if both config and credential files do not exist, create a new config file
+        if (!writer.configPath.exists() && !writer.credentialsPath.exists()) {
+            if (confirm(project, writer.configPath)) {
+                try {
+                    writer.createConfigFile()
+                } finally {
+                    AwsTelemetry.createCredentials(project)
+                }
+            } else {
+                return
```
Comment (manodnyab) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/CreateOrUpdateCredentialProfilesAction.kt:51]: Should we add a cancelled metric here?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/CreateOrUpdateCredentialProfilesAction.kt [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                message(
+                    "credentials.could_not_open",
+                    it
+                )
+            )
+        }
+
+        val fileEditorManager = FileEditorManager.getInstance(project)
+
+        localFileSystem.refreshFiles(virtualFiles, false, false) {
```
Comment (manodnyab) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/CreateOrUpdateCredentialProfilesAction.kt:68]: nit: Can you add parameter names before the false?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/CreateOrUpdateCredentialProfilesAction.kt [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            virtualFiles.forEach {
+                if (it.fileType == FileTypes.UNKNOWN) {
+                    ApplicationManager.getApplication().runWriteAction {
```
Comment (manodnyab) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/CreateOrUpdateCredentialProfilesAction.kt:71]: Is the `runWriteAction` required if we are only opening the files?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/CreateOrUpdateCredentialProfilesAction.kt (hunk 2 above)
Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/CreateOrUpdateCredentialProfilesAction.kt:28]: this file is mostly copy paste from the previous implementation

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/CreateOrUpdateCredentialProfilesAction.kt (hunk 3 above)
Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/CreateOrUpdateCredentialProfilesAction.kt:51]: this isn't really a cancellation; it's saying we didn't need to create a config file which we don't have modeled in our metrics today

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/CreateOrUpdateCredentialProfilesAction.kt (hunk 4 above)
Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/CreateOrUpdateCredentialProfilesAction.kt:68]: the linter does not like this

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/CreateOrUpdateCredentialProfilesAction.kt (hunk 5 above)
Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/CreateOrUpdateCredentialProfilesAction.kt:71]: yes

**Review by manodnyab — APPROVED** [review]

**Pull request merged**